blog/sudoko_blog_structure.md

```diff
@@ -103,42 +107,79 @@ training_data, validation_data = random_split(dataset, [0.8, 0.2], generator=gen
 
-<div align="center">
-    <img src="./threeThree.gif" height=200px />
-    <img src="./oneNine.gif" height=200px />
-    <img src="./nineOne.gif" height=200px />
-</div>
+The `MLP` model we defined above has 156,729 learnable parameters, which we will use as a baseline to compare to future models.
 
-## Loss Function
+### Training
 
-<div align="center">
-    <img src="./colouredSudoku.png" width=200px />
-    <img src="./lossVisualIdea.png" width=400px />
-</div>
+We generally trained our models on the 4m Sudoku puzzles for two epochs, since the learning of most models plateaued by this point. Later, some of our larger models required more epochs as they trained much slower, but where this happened we will mention it.
+
+We shuffled the training data at the end of each epoch to ensure that the model was seeing a different ordering of puzzles in the next epoch, and we also tested the model against validation data which it had never seen to ensure the model was not overfitting, or memorising, the training data.
+
+You can see here how our first model trained:
+
+![MLP model training loss](./images/mlp_model_training.png)
+
+### Testing our models
+
+While training the model we capture the loss which helps us train the model, but this loss number doesn't really mean much to us humans. We decided that there were two metrics that we'd be interested in also looking at, to compare model performance:
+
+1. On average, how many numbers the model accurately guessed vs. the solution
+2. What percentage of puzzles the model can solve 100% correctly (no incorrect numbers whatsoever)
+
+We planned to train our models against our standard loss calculations, and then once trained, verify the performance using these two metrics. We would run the trained model against the validation dataset to ensure it had never seen any of the puzzles before.
+
+We also read about an interesting technique to improve model performance in sudoku from a GitHub repository called [Can Convolutional Neural Networks Crack Sudoku Puzzles?](https://github.com/Kyubyong/sudoku), where they only take the single number the model is most sure about, update the input with this number and then feed this updated model back into the model, until the puzzle is completely filled out. We decided to also measure the model performance when used this way as well, which we will refer to as "iterative guesses" in the performance tables throughout this blog.
+
+So how did our `MLP` model do? Well, after being trained on over 6 million sudoku puzzles, we put it to the test and you can see the results in the table below:
```

Comment:
Does 6 million here refer to merging datasets?

Reply:
No, it's 2 epochs of 4m (since 20% is the validation set)

Reply:
I'll make that clearer